src/common/grpc_client.rs

```diff
@@ -83,47 +87,78 @@ where
         cx: &mut std::task::Context<'_>,
     ) -> std::task::Poll<Option<Self::Item>> {
         // TODO read the first 5 bytes so we know how much data to expect and we can allocate appropriately
-        let mut chunk = match self.receiver_half.poll_data(cx) {
-            Poll::Pending => return Poll::Pending,
-            Poll::Ready(r) => match r {
-                Some(r) => match r {
-                    Err(_) => return Poll::Ready(None),
-                    Ok(r) => r,
+        if self.buffer.is_empty() {
+            let chunk = match self.receiver_half.poll_data(cx) {
+                Poll::Pending => return Poll::Pending,
+                Poll::Ready(r) => match r {
+                    Some(r) => match r {
+                        Err(_) => return Poll::Ready(None),
+                        Ok(r) => r,
+                    },
+                    None => return Poll::Ready(None),
                 },
-                None => return Poll::Ready(None),
-            },
-        };
+            };
+            self.buffer = chunk;
+        }
+
+        let mut delim = self.buffer.split_to(5);
+        let _ = delim.split_to(1);
 
+        let len = u32::from_be_bytes(delim.as_ref().try_into().unwrap());
+
+        let message = self.buffer.split_to(len as usize);
         let _ = self
             .receiver_half
             .flow_control()
-            .release_capacity(chunk.len());
-        let chunk = chunk.split_off(5);
-        let p = match T::decode(chunk) {
-            Err(_) => return Poll::Pending,
+            .release_capacity(message.len() + 5);
+
+        let message = match T::decode(message) {
+            Err(e) => {
+                log::error!("decoding error {:?}", e);
+                return Poll::Pending;
+            }
             Ok(m) => m,
         };
-        Poll::Ready(Some(p))
+        Poll::Ready(Some(message))
     }
 }
 #[cfg(feature = "native")]
 type Executor<'a> = NativeExecutor<'a>;
 #[cfg(feature = "esp32")]
 type Executor<'a> = Esp32Executor<'a>;
-pub(crate) struct GrpcClient<'a> {
+pub struct GrpcClient<'a> {
     executor: Executor<'a>,
     http2_connection: SendRequest<Bytes>,
     #[allow(dead_code)]
-    http2_task: Task<()>,
+    http2_task: Option<Task<()>>,
     uri: &'a str,
 }
 
+impl<'a> Drop for GrpcClient<'a> {
```

Comment:
It's another convention. 

As these lifetime notations are local to the impl, unless you need to specify multiple lifetimes and want to add specific contexts, there's not much urgency to clarify them head of time.

This is a bit different to the elided lifetimes before because this is saying "there is some dependency between the struct and one of it's members". It can be hard to clarify that dependency with a single word. 

Sometimes they're also just there to appease the compiler.

Reply:
We could be more verbose rather than using `'a` but what it really means is that GrpcClient cannot live longer than the executor. Since GrpcClient is going to spawn tasks on the executor we need to help the compiler make sure GrpcClient lives at most as long as the Executor lives